lib/spack/spack/version/version_types.py

```diff
@@ -34,24 +33,26 @@
 class VersionStrComponent:
     __slots__ = ["data"]
 
+    data: Union[int, str]
+
     def __init__(self, data):
         # int for infinity index, str for literal.
-        self.data: Union[int, str] = data
+        self.data = data
 
     @staticmethod
-    def from_string(string):
+    def from_string(string: str) -> "VersionStrComponent":
         if len(string) >= iv_min_len:
             try:
-                string = infinity_versions.index(string)
+                return VersionStrComponent(infinity_versions.index(string))
```

Comment:
(super nitpick) The previous version was slightly better, since it was catching ValueError only from `.index` This will also cover `VersionStrComponent` initialization (which can change in the future, and maybe raise ValueError for other reasons)

Reply:
This is one of those situations that was much easier to write before type annotations. Sigh. See what you think of my new version.

Reply:
This one lgtm now, thanks!